staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -251,16 +254,90 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 			continue
 		}
 		if evalResult != types.True {
-			if len(rule.Message) != 0 {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, rule.Message))
+			if compiled.MessageExpression != nil {
+				messageExpression, newRemainingBudget, msgErr := evalMessageExpression(ctx, compiled.MessageExpression, rule.MessageExpression, activation, remainingBudget)
+				if msgErr != nil {
+					if msgErr.errType == runtimeCostErr {
+						errs = append(errs, field.InternalError(fldPath, msgErr.err))
+						return errs, -1
+					} else {
+						klog.Warning(msgErr.err)
+						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+						remainingBudget = newRemainingBudget
+					}
+				} else {
+					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					remainingBudget = newRemainingBudget
+				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, fmt.Sprintf("failed rule: %s", ruleErrorString(rule))))
+				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
 			}
 		}
 	}
 	return errs, remainingBudget
 }
 
+type _errType int
```

Comment:
nit: Use idiomatic typed errors here?

```go
type expressionError struct{...}
func (*expressionError) Error() string { 
...
}
```
`evalMessageExpression()` can then return `error` and callers can use `errors.As` ?

Reply:
Note that you probably only need a typed error for runtime cost errors since you could then return all eval errors as normal errors.

Reply:
Right, makes sense, I'll refactor accordingly.